flink-libraries/flink-state-processing-api/src/main/java/org/apache/flink/state/api/input/KeyedStateInputFormat.java

```diff
@@ -0,0 +1,253 @@
+
+	private static final String USER_TIMERS_NAME = "user-timers";
+
+	private final StateBackend stateBackend;
+
+	private final TypeInformation<K> keyType;
+
+	private final KeyedStateReaderFunction<K, OUT> userFunction;
+
+	private transient TypeSerializer<K> keySerializer;
+
+	private transient CloseableRegistry registry;
+
+	private transient BufferingCollector<OUT> out;
+
+	private transient Iterator<K> keys;
+
+	private transient AbstractKeyedStateBackend<K> keyedStateBackend;
+
+	private transient Context ctx;
+
+	/**
+	 * Creates an input format for reading partitioned state from an operator in a savepoint.
+	 *
+	 * @param savepointPath The path to an existing savepoint.
+	 * @param uid           The uid of an operator.
+	 * @param stateBackend  The state backed used to snapshot the operator.
+	 * @param keyType       The type information describing the key type.
+	 * @param userFunction  The {@link KeyedStateReaderFunction} called for each key in the operator.
+	 */
+	public KeyedStateInputFormat(
+		String savepointPath,
+		String uid,
+		StateBackend stateBackend,
+		TypeInformation<K> keyType,
+		KeyedStateReaderFunction<K, OUT> userFunction) {
+		super(savepointPath, uid);
+		this.stateBackend = stateBackend;
+		this.keyType = keyType;
+		this.userFunction = userFunction;
+	}
+
+	@Override
+	public KeyGroupRangeInputSplit[] createInputSplits(int minNumSplits) throws IOException {
+		final OperatorState operatorState = getOperatorState();
+
+		return getKeyGroupRangeInputSplits(minNumSplits, operatorState);
+	}
+
+	@Override
+	public void openInputFormat() {
+		out = new BufferingCollector<>();
+		keySerializer = keyType.createSerializer(getRuntimeContext().getExecutionConfig());
+	}
+
+	@Override
+	@SuppressWarnings("unchecked")
+	public void open(KeyGroupRangeInputSplit split) throws IOException {
+		registry = new CloseableRegistry();
+
+		final Environment environment = new SavepointEnvironment
+			.Builder(getRuntimeContext(), split.getNumKeyGroups())
+			.setSubtaskIndex(split.getSplitNumber())
+			.setPrioritizedOperatorSubtaskState(split.getPrioritizedOperatorSubtaskState())
+			.build();
+
+		final StreamOperatorStateContext context = getStreamOperatorStateContext(environment);
+
+		keyedStateBackend = (AbstractKeyedStateBackend<K>) context.keyedStateBackend();
+		keys = getKeyIterator();
+		ctx = new Context();
+	}
+
+	@SuppressWarnings("unchecked")
+	private Iterator<K> getKeyIterator() throws IOException {
+		Preconditions.checkNotNull(keyedStateBackend);
+		final DefaultKeyedStateStore keyedStateStore = new DefaultKeyedStateStore(keyedStateBackend, getRuntimeContext().getExecutionConfig());
+
+		final List<StateDescriptor<?, ?>> stateDescriptors;
+		try (SavepointRuntimeContext ctx = new SavepointRuntimeContext(getRuntimeContext(), keyedStateStore)) {
+			FunctionUtils.setFunctionRuntimeContext(userFunction, ctx);
```

Comment:
I feel like we should be setting the runtime context already in the `open()` method, instead of here, since the concern of this method should really only be about getting the key iterator.